Add configuration to skip in Maven plugin

In multi-module projects usually the parent POM is used to specify plugin and dependency versions to be used in child modules. Then, there are build plugins that are not desired to be enabled for all modules. Therefore it has become a common practice by Maven plugin to support a 'skip' configuration item. If skip = true then the plugin should skip that module. Then we can for example set skip = true as default and enable it (skip = false) for modules that need that plugin.

So it would be awesome if the jib-maven-plugin also could support such a 'skip' configuration.

Sorry if I was not clear in my original message. It currently does not work
to simply add the plugin to pluginManagement in the parent-pom and add the
plugin (without version etc) to modules that need it. In this case it fails
with the following error message:
"[ERROR] Failed to execute goal
com.google.cloud.tools:jib-maven-plugin:0.9.9:build (default-cli) on
project my-library: Missing target image parameter, perhaps you should add
a <to><image> configuration parameter to your pom.xml or set the parameter
via the commandline (e.g. 'mvn compile jib:build -Dimage=<your image
name>'). -> [Help 1]

In this case my project is a parent-pom with two modules, library and app.
The app module depends on the library module and is a Spring Boot
application. In the parent-pom pluginManagement I've added the Jib plugin
with version and default configuration (like base image and container
args). Then in the app-module I've added executions for the Jib plugin.

As can be seen from the Maven error message it will try to execute the
plugin for the library-module now because there is no way to tell Maven to
skip the plugin for that module. The parent-pom is skipped but that is
because Jib always skips pom modules.

BR, Andreas



On Tue, Aug 28, 2018 at 3:58 AM Appu <notifications@github.com> wrote:

> So you can configure it in the parent and use it only on the children,
> using pluginManagement:
> https://maven.apache.org/pom.html#Plugin_Management. You can still do
> common configuration in the parent as you would like, and you wouldn't have
> to reference the plugin at all on modules that don't use it.
>
> If the simple case that you are running mvn package, you bind the jib
> goal to the package phase (you can even go as far as to unbind the jar/war
> goal) and it all works the same.
>
> Perhaps *hack* is overly aggressive in this situation, but applying the
> plugin everywhere and then disabling it selectively seems like a less
> meaningful pattern than only enabling it when necessary.
>
> Perhaps I'm missing something though?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416425412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAz1gLnbFTIzagbB8budB1Q_NysOkstbks5uVKOxgaJpZM4WHq6j>
> .
>

Do you happen to have the plugin also defined in the <build> tag of your
parent pom?

On Tue, Aug 28, 2018 at 6:19 PM Andreas Häber <notifications@github.com>
wrote:

> Sorry if I was not clear in my original message. It currently does not work
> to simply add the plugin to pluginManagement in the parent-pom and add the
> plugin (without version etc) to modules that need it. In this case it fails
> with the following error message:
> "[ERROR] Failed to execute goal
> com.google.cloud.tools:jib-maven-plugin:0.9.9:build (default-cli) on
> project my-library: Missing target image parameter, perhaps you should add
> a <to><image> configuration parameter to your pom.xml or set the parameter
> via the commandline (e.g. 'mvn compile jib:build -Dimage=<your image
> name>'). -> [Help 1]
>
> In this case my project is a parent-pom with two modules, library and app.
> The app module depends on the library module and is a Spring Boot
> application. In the parent-pom pluginManagement I've added the Jib plugin
> with version and default configuration (like base image and container
> args). Then in the app-module I've added executions for the Jib plugin.
>
> As can be seen from the Maven error message it will try to execute the
> plugin for the library-module now because there is no way to tell Maven to
> skip the plugin for that module. The parent-pom is skipped but that is
> because Jib always skips pom modules.
>
> BR, Andreas
>
>
>
> On Tue, Aug 28, 2018 at 3:58 AM Appu <notifications@github.com> wrote:
>
> > So you can configure it in the parent and use it only on the children,
> > using pluginManagement:
> > https://maven.apache.org/pom.html#Plugin_Management. You can still do
> > common configuration in the parent as you would like, and you wouldn't
> have
> > to reference the plugin at all on modules that don't use it.
> >
> > If the simple case that you are running mvn package, you bind the jib
> > goal to the package phase (you can even go as far as to unbind the
> jar/war
> > goal) and it all works the same.
> >
> > Perhaps *hack* is overly aggressive in this situation, but applying the
> > plugin everywhere and then disabling it selectively seems like a less
> > meaningful pattern than only enabling it when necessary.
> >
> > Perhaps I'm missing something though?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416425412
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAz1gLnbFTIzagbB8budB1Q_NysOkstbks5uVKOxgaJpZM4WHq6j
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416530823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHoJzY1q5Ic9XPvHtPedSvH4FgRG0Cfuks5uVRlIgaJpZM4WHq6j>
> .
>

No, only in pluginManagement. The command I use is:
mvn package jib:build

BR, Andreas

On Tue, Aug 28, 2018 at 1:06 PM Darth Binamira <notifications@github.com>
wrote:

> Do you happen to have the plugin also defined in the <build> tag of your
> parent pom?
>
> On Tue, Aug 28, 2018 at 6:19 PM Andreas Häber <notifications@github.com>
> wrote:
>
> > Sorry if I was not clear in my original message. It currently does not
> work
> > to simply add the plugin to pluginManagement in the parent-pom and add
> the
> > plugin (without version etc) to modules that need it. In this case it
> fails
> > with the following error message:
> > "[ERROR] Failed to execute goal
> > com.google.cloud.tools:jib-maven-plugin:0.9.9:build (default-cli) on
> > project my-library: Missing target image parameter, perhaps you should
> add
> > a <to><image> configuration parameter to your pom.xml or set the
> parameter
> > via the commandline (e.g. 'mvn compile jib:build -Dimage=<your image
> > name>'). -> [Help 1]
> >
> > In this case my project is a parent-pom with two modules, library and
> app.
> > The app module depends on the library module and is a Spring Boot
> > application. In the parent-pom pluginManagement I've added the Jib plugin
> > with version and default configuration (like base image and container
> > args). Then in the app-module I've added executions for the Jib plugin.
> >
> > As can be seen from the Maven error message it will try to execute the
> > plugin for the library-module now because there is no way to tell Maven
> to
> > skip the plugin for that module. The parent-pom is skipped but that is
> > because Jib always skips pom modules.
> >
> > BR, Andreas
> >
> >
> >
> > On Tue, Aug 28, 2018 at 3:58 AM Appu <notifications@github.com> wrote:
> >
> > > So you can configure it in the parent and use it only on the children,
> > > using pluginManagement:
> > > https://maven.apache.org/pom.html#Plugin_Management. You can still do
> > > common configuration in the parent as you would like, and you wouldn't
> > have
> > > to reference the plugin at all on modules that don't use it.
> > >
> > > If the simple case that you are running mvn package, you bind the jib
> > > goal to the package phase (you can even go as far as to unbind the
> > jar/war
> > > goal) and it all works the same.
> > >
> > > Perhaps *hack* is overly aggressive in this situation, but applying the
> > > plugin everywhere and then disabling it selectively seems like a less
> > > meaningful pattern than only enabling it when necessary.
> > >
> > > Perhaps I'm missing something though?
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416425412
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AAz1gLnbFTIzagbB8budB1Q_NysOkstbks5uVKOxgaJpZM4WHq6j
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416530823
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AHoJzY1q5Ic9XPvHtPedSvH4FgRG0Cfuks5uVRlIgaJpZM4WHq6j
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleContainerTools/jib/issues/865#issuecomment-416544918>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAz1gLEyvNbMkFOymRH2mPXq7oXl9qUyks5uVSRFgaJpZM4WHq6j>
> .
>

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
@@ -43,6 +43,10 @@ public class BuildDockerMojo extends JibPluginConfiguration {
 
   @Override
   public void execute() throws MojoExecutionException {
+    if (isSkipped()) {
+      getLog().info("Skipping containerizations because jib-maven-plugin: skip = true");
+      return;
+    }
     if ("pom".equals(getProject().getPackaging())) {
       getLog().info("Skipping containerization because packaging is 'pom'...");
       return;

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
@@ -44,7 +44,7 @@ public class BuildDockerMojo extends JibPluginConfiguration {
   @Override
   public void execute() throws MojoExecutionException {
     if (isSkipped()) {
-      getLog().info("Skipping containerizations because jib-maven-plugin: skip = true");
+      getLog().info("Skipping containerization because jib-maven-plugin: skip = true");
       return;
     }
     if ("pom".equals(getProject().getPackaging())) {

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
@@ -48,6 +48,10 @@ public class BuildImageMojo extends JibPluginConfiguration {
 
   @Override
   public void execute() throws MojoExecutionException, MojoFailureException {
+    if (isSkipped()) {
+      getLog().info("Skipping containerization because jib-maven-plugin: skip = true");
+      return;
+    }
     if ("pom".equals(getProject().getPackaging())) {
       getLog().info("Skipping containerization because packaging is 'pom'...");
       return;

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
@@ -45,6 +45,10 @@ public class BuildTarMojo extends JibPluginConfiguration {
 
   @Override
   public void execute() throws MojoExecutionException {
+    if (isSkipped()) {
+      getLog().info("Skipping containerization because jib-maven-plugin: skip = true");
+      return;
+    }
     if ("pom".equals(getProject().getPackaging())) {
       getLog().info("Skipping containerization because packaging is 'pom'...");
       return;

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/DockerContextMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/DockerContextMojo.java
@@ -50,6 +50,10 @@ public class DockerContextMojo extends JibPluginConfiguration {
 
   @Override
   public void execute() throws MojoExecutionException {
+    if (isSkipped()) {
+      getLog().info("Skipping containerization because jib-maven-plugin: skip = true");
+      return;
+    }
     if ("pom".equals(getProject().getPackaging())) {
       getLog().info("Skipping containerization because packaging is 'pom'...");
       return;

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java
@@ -170,6 +170,9 @@ abstract class JibPluginConfiguration extends AbstractMojo {
   @Parameter(defaultValue = "${project.basedir}/src/main/jib", required = true)
   private String extraDirectory;
 
+  @Parameter(defaultValue = "false", property = "jib.skip")
+  private boolean skip;
+
   @Nullable @Component protected SettingsDecrypter settingsDecrypter;
 
   /** Default constructor handles setting up auth property descriptors. */
@@ -300,6 +303,10 @@ abstract class JibPluginConfiguration extends AbstractMojo {
   Path getExtraDirectory() {
     // TODO: Should inform user about nonexistent directory if using custom directory.
     return Paths.get(Preconditions.checkNotNull(extraDirectory));
+  }
+
+  boolean isSkipped() {
+    return skip;
   }
 
   SettingsDecrypter getSettingsDecrypter() {

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
@@ -17,6 +17,9 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
+import com.google.common.base.Charsets;
+import com.google.common.io.Files;
+import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.time.Instant;
@@ -41,6 +44,9 @@ public class BuildDockerMojoIntegrationTest {
   @ClassRule
   public static final TestProject defaultTargetTestProject =
       new TestProject(testPlugin, "default-target");
+
+  @ClassRule
+  public static final TestProject skippedTestProject = new TestProject(testPlugin, "empty");
 
   /**
    * Builds and runs jib:buildDocker on a project at {@code projectRoot} pushing to {@code
@@ -109,5 +115,25 @@ public class BuildDockerMojoIntegrationTest {
         buildToDockerDaemonAndRun(
             defaultTargetTestProject.getProjectRoot(),
             "default-target-name:default-target-version"));
+  }
+
+  @Test
+  public void testExecute_skipJibGoal() throws VerificationException, IOException {
+    String targetImage = "emptyimage:maven" + System.nanoTime();
+
+    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
+    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
+    verifier.setAutoclean(false);
+    verifier.setSystemProperty("jib.skip", "true");
+
+    verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
+    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertTrue(
+        Files.asCharSource(logFile, Charsets.UTF_8)
+            .read()
+            .contains(
+                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
+                    + "[INFO] ------------------------------------------------------------------------\n"
+                    + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
@@ -18,10 +18,7 @@ package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
-import java.nio.charset.Charset;
-import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.Paths;
 import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
@@ -119,21 +116,6 @@ public class BuildDockerMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "neverbuilt:maven" + System.nanoTime();
-
-    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
-    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
-    verifier.setAutoclean(false);
-    verifier.setSystemProperty("jib.skip", "true");
-
-    verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
-
-    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertThat(
-        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
-        CoreMatchers.containsString(
-            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
-                + "[INFO] ------------------------------------------------------------------------\n"
-                + "[INFO] BUILD SUCCESS"));
+    SkippedGoalVerifier.verifyGoalIsSkipped(skippedTestProject, BuildDockerMojo.GOAL_NAME);
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
@@ -17,11 +17,11 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
-import com.google.common.base.Charsets;
-import com.google.common.io.Files;
-import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
@@ -119,7 +119,7 @@ public class BuildDockerMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "emptyimage:maven" + System.nanoTime();
+    String targetImage = "neverbuilt:maven" + System.nanoTime();
 
     Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
     verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
@@ -127,13 +127,13 @@ public class BuildDockerMojoIntegrationTest {
     verifier.setSystemProperty("jib.skip", "true");
 
     verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
-    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertTrue(
-        Files.asCharSource(logFile, Charsets.UTF_8)
-            .read()
-            .contains(
-                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
-                    + "[INFO] ------------------------------------------------------------------------\n"
-                    + "[INFO] BUILD SUCCESS"));
+
+    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertThat(
+        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        CoreMatchers.containsString(
+            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
+                + "[INFO] ------------------------------------------------------------------------\n"
+                + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
@@ -19,6 +19,8 @@ package com.google.cloud.tools.jib.maven;
 import com.google.cloud.tools.jib.Command;
 import com.google.cloud.tools.jib.IntegrationTestingConfiguration;
 import com.google.cloud.tools.jib.registry.LocalRegistry;
+import com.google.common.base.Charsets;
+import java.io.File;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -58,6 +60,9 @@ public class BuildImageMojoIntegrationTest {
 
   @ClassRule
   public static final TestProject emptyTestProject = new TestProject(testPlugin, "empty");
+
+  @ClassRule
+  public static final TestProject skippedTestProject = new TestProject(testPlugin, "empty");
 
   @ClassRule
   public static final TestProject defaultTargetTestProject =
@@ -269,5 +274,25 @@ public class BuildImageMojoIntegrationTest {
     Assert.assertEquals(
         "Hello, world. An argument.\nfoo\ncat\n-Xms512m\n-Xdebug\n",
         buildAndRunComplex(targetImage, "testuser", "testpassword", localRegistry1));
+  }
+
+  @Test
+  public void testExecute_skipJibGoal() throws VerificationException, IOException {
+    String targetImage = "emptyimage:maven" + System.nanoTime();
+
+    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
+    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
+    verifier.setAutoclean(false);
+    verifier.setSystemProperty("jib.skip", "true");
+
+    verifier.executeGoal("jib:" + BuildImageMojo.GOAL_NAME);
+    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertTrue(
+        com.google.common.io.Files.asCharSource(logFile, Charsets.UTF_8)
+            .read()
+            .contains(
+                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
+                    + "[INFO] ------------------------------------------------------------------------\n"
+                    + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
@@ -20,7 +20,6 @@ import com.google.cloud.tools.jib.Command;
 import com.google.cloud.tools.jib.IntegrationTestingConfiguration;
 import com.google.cloud.tools.jib.registry.LocalRegistry;
 import java.io.IOException;
-import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -277,21 +276,6 @@ public class BuildImageMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "neverbuilt:maven" + System.nanoTime();
-
-    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
-    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
-    verifier.setAutoclean(false);
-    verifier.setSystemProperty("jib.skip", "true");
-
-    verifier.executeGoal("jib:" + BuildImageMojo.GOAL_NAME);
-
-    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertThat(
-        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
-        CoreMatchers.containsString(
-            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
-                + "[INFO] ------------------------------------------------------------------------\n"
-                + "[INFO] BUILD SUCCESS"));
+    SkippedGoalVerifier.verifyGoalIsSkipped(skippedTestProject, BuildImageMojo.GOAL_NAME);
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
@@ -19,9 +19,8 @@ package com.google.cloud.tools.jib.maven;
 import com.google.cloud.tools.jib.Command;
 import com.google.cloud.tools.jib.IntegrationTestingConfiguration;
 import com.google.cloud.tools.jib.registry.LocalRegistry;
-import com.google.common.base.Charsets;
-import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -278,7 +277,7 @@ public class BuildImageMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "emptyimage:maven" + System.nanoTime();
+    String targetImage = "neverbuilt:maven" + System.nanoTime();
 
     Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
     verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
@@ -286,13 +285,13 @@ public class BuildImageMojoIntegrationTest {
     verifier.setSystemProperty("jib.skip", "true");
 
     verifier.executeGoal("jib:" + BuildImageMojo.GOAL_NAME);
-    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertTrue(
-        com.google.common.io.Files.asCharSource(logFile, Charsets.UTF_8)
-            .read()
-            .contains(
-                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
-                    + "[INFO] ------------------------------------------------------------------------\n"
-                    + "[INFO] BUILD SUCCESS"));
+
+    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertThat(
+        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        CoreMatchers.containsString(
+            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
+                + "[INFO] ------------------------------------------------------------------------\n"
+                + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
@@ -17,6 +17,9 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
+import com.google.common.base.Charsets;
+import com.google.common.io.Files;
+import java.io.File;
 import java.io.IOException;
 import java.time.Instant;
 import org.apache.maven.it.VerificationException;
@@ -31,6 +34,9 @@ public class BuildTarMojoIntegrationTest {
 
   @ClassRule
   public static final TestProject simpleTestProject = new TestProject(testPlugin, "simple");
+
+  @ClassRule
+  public static final TestProject skippedTestProject = new TestProject(testPlugin, "empty");
 
   /**
    * Builds and runs jib:buildTar on a project at {@code projectRoot} pushing to {@code
@@ -66,5 +72,25 @@ public class BuildTarMojoIntegrationTest {
         Instant.parse(
             new Command("docker", "inspect", "-f", "{{.Created}}", targetImage).run().trim());
     Assert.assertTrue(buildTime.isAfter(before) || buildTime.equals(before));
+  }
+
+  @Test
+  public void testExecute_skipJibGoal() throws VerificationException, IOException {
+    String targetImage = "emptyimage:maven" + System.nanoTime();
+
+    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
+    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
+    verifier.setAutoclean(false);
+    verifier.setSystemProperty("jib.skip", "true");
+
+    verifier.executeGoal("jib:" + BuildTarMojo.GOAL_NAME);
+    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertTrue(
+        Files.asCharSource(logFile, Charsets.UTF_8)
+            .read()
+            .contains(
+                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
+                    + "[INFO] ------------------------------------------------------------------------\n"
+                    + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
@@ -17,13 +17,15 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
-import com.google.common.base.Charsets;
-import com.google.common.io.Files;
-import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
+import org.hamcrest.CoreMatchers;
 import org.junit.Assert;
 import org.junit.ClassRule;
 import org.junit.Test;
@@ -76,7 +78,7 @@ public class BuildTarMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "emptyimage:maven" + System.nanoTime();
+    String targetImage = "neverbuilt:maven" + System.nanoTime();
 
     Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
     verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
@@ -84,13 +86,13 @@ public class BuildTarMojoIntegrationTest {
     verifier.setSystemProperty("jib.skip", "true");
 
     verifier.executeGoal("jib:" + BuildTarMojo.GOAL_NAME);
-    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertTrue(
-        Files.asCharSource(logFile, Charsets.UTF_8)
-            .read()
-            .contains(
-                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
-                    + "[INFO] ------------------------------------------------------------------------\n"
-                    + "[INFO] BUILD SUCCESS"));
+
+    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertThat(
+        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        CoreMatchers.containsString(
+            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
+                + "[INFO] ------------------------------------------------------------------------\n"
+                + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
@@ -18,14 +18,9 @@ package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
-import java.nio.charset.Charset;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
 import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
-import org.hamcrest.CoreMatchers;
 import org.junit.Assert;
 import org.junit.ClassRule;
 import org.junit.Test;
@@ -78,21 +73,6 @@ public class BuildTarMojoIntegrationTest {
 
   @Test
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "neverbuilt:maven" + System.nanoTime();
-
-    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
-    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
-    verifier.setAutoclean(false);
-    verifier.setSystemProperty("jib.skip", "true");
-
-    verifier.executeGoal("jib:" + BuildTarMojo.GOAL_NAME);
-
-    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertThat(
-        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
-        CoreMatchers.containsString(
-            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
-                + "[INFO] ------------------------------------------------------------------------\n"
-                + "[INFO] BUILD SUCCESS"));
+    SkippedGoalVerifier.verifyGoalIsSkipped(skippedTestProject, BuildTarMojo.GOAL_NAME);
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
@@ -18,10 +18,8 @@ package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
-import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.Paths;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
 import org.hamcrest.CoreMatchers;
@@ -79,21 +77,6 @@ public class DockerContextMojoIntegrationTest {
   }
 
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "neverbuilt:maven" + System.nanoTime();
-
-    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
-    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
-    verifier.setAutoclean(false);
-    verifier.setSystemProperty("jib.skip", "true");
-
-    verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
-
-    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertThat(
-        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
-        CoreMatchers.containsString(
-            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
-                + "[INFO] ------------------------------------------------------------------------\n"
-                + "[INFO] BUILD SUCCESS"));
+    SkippedGoalVerifier.verifyGoalIsSkipped(skippedTestProject, BuildDockerMojo.GOAL_NAME);
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
@@ -17,11 +17,11 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
-import com.google.common.base.Charsets;
-import java.io.File;
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
 import org.hamcrest.CoreMatchers;
@@ -79,7 +79,7 @@ public class DockerContextMojoIntegrationTest {
   }
 
   public void testExecute_skipJibGoal() throws VerificationException, IOException {
-    String targetImage = "emptyimage:maven" + System.nanoTime();
+    String targetImage = "neverbuilt:maven" + System.nanoTime();
 
     Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
     verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
@@ -87,13 +87,13 @@ public class DockerContextMojoIntegrationTest {
     verifier.setSystemProperty("jib.skip", "true");
 
     verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
-    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
-    Assert.assertTrue(
-        com.google.common.io.Files.asCharSource(logFile, Charsets.UTF_8)
-            .read()
-            .contains(
-                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
-                    + "[INFO] ------------------------------------------------------------------------\n"
-                    + "[INFO] BUILD SUCCESS"));
+
+    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertThat(
+        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        CoreMatchers.containsString(
+            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
+                + "[INFO] ------------------------------------------------------------------------\n"
+                + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/DockerContextMojoIntegrationTest.java
@@ -17,6 +17,8 @@
 package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
+import com.google.common.base.Charsets;
+import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -37,6 +39,9 @@ public class DockerContextMojoIntegrationTest {
 
   @ClassRule
   public static final TestProject simpleTestProject = new TestProject(testPlugin, "simple");
+
+  @ClassRule
+  public static final TestProject skippedTestProject = new TestProject(testPlugin, "empty");
 
   @Test
   public void testExecute() throws VerificationException, IOException, InterruptedException {
@@ -71,5 +76,24 @@ public class DockerContextMojoIntegrationTest {
 
     Assert.assertEquals(
         "Hello, world. An argument.\nfoo\ncat\n", new Command("docker", "run", imageName).run());
+  }
+
+  public void testExecute_skipJibGoal() throws VerificationException, IOException {
+    String targetImage = "emptyimage:maven" + System.nanoTime();
+
+    Verifier verifier = new Verifier(skippedTestProject.getProjectRoot().toString());
+    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
+    verifier.setAutoclean(false);
+    verifier.setSystemProperty("jib.skip", "true");
+
+    verifier.executeGoal("jib:" + BuildDockerMojo.GOAL_NAME);
+    File logFile = new File(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertTrue(
+        com.google.common.io.Files.asCharSource(logFile, Charsets.UTF_8)
+            .read()
+            .contains(
+                "[INFO] Skipping containerizations because jib-maven-plugin: skip = true\n"
+                    + "[INFO] ------------------------------------------------------------------------\n"
+                    + "[INFO] BUILD SUCCESS"));
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Google LLC. All rights reserved.
+ * Copyright 2018 Google LLC.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not
  * use this file except in compliance with the License. You may obtain a copy of

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
@@ -1,0 +1,34 @@
+package com.google.cloud.tools.jib.maven;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.apache.maven.it.VerificationException;
+import org.apache.maven.it.Verifier;
+import org.hamcrest.CoreMatchers;
+import org.junit.Assert;
+
+public class SkippedGoalVerifier {
+
+  public static void verifyGoalIsSkipped(TestProject testProject, String goal)
+      throws VerificationException, IOException {
+    String targetImage = "neverbuilt:maven" + System.nanoTime();
+
+    Verifier verifier = new Verifier(testProject.getProjectRoot().toString());
+    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
+    verifier.setAutoclean(false);
+    verifier.setSystemProperty("jib.skip", "true");
+
+    verifier.executeGoal("jib:" + goal);
+
+    Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
+    Assert.assertThat(
+        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        CoreMatchers.containsString(
+            "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
+                + "[INFO] ------------------------------------------------------------------------\n"
+                + "[INFO] BUILD SUCCESS"));
+  }
+}

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
@@ -1,7 +1,23 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
 package com.google.cloud.tools.jib.maven;
 
 import java.io.IOException;
-import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -10,14 +26,16 @@ import org.apache.maven.it.Verifier;
 import org.hamcrest.CoreMatchers;
 import org.junit.Assert;
 
-public class SkippedGoalVerifier {
+/** A simple verifier utility to test goal skipping accross all our jib goals. */
+class SkippedGoalVerifier {
 
-  public static void verifyGoalIsSkipped(TestProject testProject, String goal)
+  private SkippedGoalVerifier() {}
+
+  /** Verify that a jib goal is skipped */
+  static void verifyGoalIsSkipped(TestProject testProject, String goal)
       throws VerificationException, IOException {
-    String targetImage = "neverbuilt:maven" + System.nanoTime();
 
     Verifier verifier = new Verifier(testProject.getProjectRoot().toString());
-    verifier.setSystemProperty("_TARGET_IMAGE", targetImage);
     verifier.setAutoclean(false);
     verifier.setSystemProperty("jib.skip", "true");
 
@@ -25,7 +43,7 @@ public class SkippedGoalVerifier {
 
     Path logFile = Paths.get(verifier.getBasedir(), verifier.getLogFileName());
     Assert.assertThat(
-        new String(Files.readAllBytes(logFile), Charset.forName("UTF-8")),
+        new String(Files.readAllBytes(logFile), StandardCharsets.UTF_8),
         CoreMatchers.containsString(
             "[INFO] Skipping containerization because jib-maven-plugin: skip = true\n"
                 + "[INFO] ------------------------------------------------------------------------\n"

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/SkippedGoalVerifier.java
@@ -29,12 +29,9 @@ import org.junit.Assert;
 /** A simple verifier utility to test goal skipping accross all our jib goals. */
 class SkippedGoalVerifier {
 
-  private SkippedGoalVerifier() {}
-
   /** Verify that a jib goal is skipped */
   static void verifyGoalIsSkipped(TestProject testProject, String goal)
       throws VerificationException, IOException {
-
     Verifier verifier = new Verifier(testProject.getProjectRoot().toString());
     verifier.setAutoclean(false);
     verifier.setSystemProperty("jib.skip", "true");
@@ -49,4 +46,6 @@ class SkippedGoalVerifier {
                 + "[INFO] ------------------------------------------------------------------------\n"
                 + "[INFO] BUILD SUCCESS"));
   }
+
+  private SkippedGoalVerifier() {}
 }

